apps/projects/app/App.js

```diff
@@ -61,40 +62,39 @@ const App = () => {
     setSelectedIssue(selectedIssueId)
   }, [selectedIssueId])
 
-  useEffect(() => {
-    const code = getURLParam('code')
-    code &&
-      window.opener.postMessage(
-        { from: 'popup', name: 'code', value: code },
-        '*'
-      )
-    window.close()
-  })
-
-  const handlePopupMessage = async message => {
-    if (message.data.from !== 'popup') return
-    if (message.data.name === 'code') {
-      // TODO: Optimize the listeners lifecycle, ie: remove on unmount
-      window.removeEventListener('message', handlePopupMessage)
+  const handlePopupMessage = useCallback(message => {
+    if (!popupRef) return
+    if (message.source !== popupRef) return
+
+    popupRef = null
 
-      const code = message.data.value
+    async function processCode() {
```

Comment:
Why define a function inside the callback, using variables from the outer scope (`message`) just call that function from the callback context? Shouldn't be enough to put the logic directly in the callback?
Is it related to the try-catch? Also the function is being declared `async` but the call is `sync` the general structure of this callback confuses me a bit.

Reply:
A function passed to `useCallback` cannot return anything (except another function, for the use-case of removing an event listener)

An `async` function returns a promise – that's part of the behind-the-scenes magic of `async/await`

So if you want to use `async/await` within a `useCallback`, you have to define it inline like this and then call it. We are doing this in a few spots in our code now.

An alternative would be using regular Promise notation:

```js
const handlePopupMessage = useCallback(message => {
  if (!popupRef) return
  if (message.source !== popupRef) return

  popupRef = null

  getToken(message.data.value)
    .then(token => {
      setGithubLoading(false)
      api.emitTrigger(REQUESTED_GITHUB_TOKEN_SUCCESS, {
        status: STATUS.AUTHENTICATED,
        token
      })
    })
    .catch(err => {
      console.error(err)
      setGithubLoading(false)
      api.emitTrigger(REQUESTED_GITHUB_TOKEN_FAILURE, {
        status: STATUS.FAILED,
        token: null,
      })
    })
}, [api])
```

Do you like this more, @ottodevs?

Reply:
I think it is enough to have it the way you did, as it seems consistent with other similar patterns in our codebase. I really appreciate your explanation about this change! Thank you!